File: sdks/remote_plugin/python/app/core/tester.py
from __future__ import annotations

import asyncio
from collections.abc import Callable
from typing import Any

from app.core.flow import Flow
from app.core.plugin import AsyncPlugin
from app.schema.payload import Payload
from app.schema.signal import Signal

ACTIVITY_BUFFER_SIZE = 1024


class MockFlow(Flow):
    """MockFlow implements the Flow interface hooks for testing purposes."""

    def __init__(self, ping_activity: Callable[[], None] | None = None) -> None:
        self._ping_activity = ping_activity
        self.on_signal_hook: Callable[[int, Signal], None] | None = None
        self.on_payload_hook: Callable[[int, Payload], None] | None = None

    def ping(self) -> None:
        if self._ping_activity:
            self._ping_activity()

    async def send_payload(
        self,
        payload: Payload,
        *,
        port: int = 0,
    ) -> None:
        self.ping()
        if self.on_payload_hook:
            self.on_payload_hook(port, payload)

    async def send_signal(
        self,
        signal: Signal,
        *,
        port: int = 0,
    ) -> None:
        self.ping()
        if self.on_signal_hook:
            self.on_signal_hook(port, signal)


class PluginTester:
    """PluginTester provides a harness for testing individual plugins in isolation.

    It uses a MockFlow to capture outputs and track plugin activity.
    """

    def __init__(self, plugin: AsyncPlugin[Any]) -> None:
        self.plugin = plugin
        self.mock = MockFlow(self._ping_activity)
        self._activity: asyncio.Queue[None] = asyncio.Queue(maxsize=ACTIVITY_BUFFER_SIZE)

    def _ping_activity(self) -> None:
        try:
            self._activity.put_nowait(None)
        except asyncio.QueueFull:
            pass

    async def start(self) -> PluginTester:
        await self.plugin.on_start(self.mock)
        await self.plugin.on_ready(self.mock)
        return self

    async def stop(self) -> None:
        await self.plugin.on_stop()

    async def wait(self, timeout_sec: float = 0.1) -> None:
        """Wait blocks until the plugin stops emitting activity for the specified duration.

        This is useful for testing asynchronous plugins that process data in the background.
        """
        while True:
            try:
                await asyncio.wait_for(self._activity.get(), timeout=timeout_sec)
                self._activity.task_done()
            except TimeoutError:
                # No activity for `timeout` seconds, we consider it done.
                return

    async def inject_signal(self, signal: Signal) -> PluginTester:
        self._ping_activity()
        await self.plugin.on_signal(self.mock, signal)
        return self

    async def inject_payload(self, payload: Payload) -> PluginTester:
        self._ping_activity()
        await self.plugin.on_payload(self.mock, payload)
        return self

    def on_signal(self, cb: Callable[[int, Signal], None]) -> PluginTester:
        self.mock.on_signal_hook = cb
        return self

    def on_payload(self, cb: Callable[[int, Payload], None]) -> PluginTester:
        self.mock.on_payload_hook = cb
        return self

File: sdks/remote_plugin/nodejs/src/core/tester.ts
import type { Flow } from './flow.js'
import type { AsyncPlugin } from './plugin.js'
import { Signal } from '../schema/signal.js'
import { Payload } from '../schema/payload.js'

const ACTIVITY_BUFFER_SIZE = 1024

/**
 * MockFlow implements the Flow interface for testing purposes.
 * Captures emitted payloads and signals via hooks.
 */
export class MockFlow implements Flow {
  private readonly pingActivity?: () => void

  onSignalHook?: (port: number, signal: Signal) => void
  onPayloadHook?: (port: number, payload: Payload) => void

  constructor(pingActivity?: () => void) {
    this.pingActivity = pingActivity
  }

  async sendPayload(payload: Payload, options?: { port?: number }): Promise<void> {
    this.pingActivity?.()
    this.onPayloadHook?.(options?.port ?? 0, payload)
  }

  async sendSignal(signal: Signal, options?: { port?: number }): Promise<void> {
    this.pingActivity?.()
    this.onSignalHook?.(options?.port ?? 0, signal)
  }
}

/**
 * PluginTester provides a test harness for individual plugins.
 *
 * @example
 * ```ts
 * const tester = new PluginTester(new MyPlugin(config))
 * tester.onPayload((port, payload) => { ... })
 * await tester.start()
 * await tester.injectPayload(new Payload('test', { key: 'val' }))
 * await tester.wait()
 * await tester.stop()
 * ```
 */
export class PluginTester {
  readonly plugin: AsyncPlugin<unknown>
  readonly mock: MockFlow

  private activityBuffer: Array<() => void> = []
  private activityWaiter: (() => void) | null = null

  constructor(plugin: AsyncPlugin<unknown>) {
    this.plugin = plugin
    this.mock = new MockFlow(() => this.pingActivity())
  }

  async start(): Promise<this> {
    await this.plugin.onStart(this.mock)
    await this.plugin.onReady(this.mock)
    return this
  }

  async stop(): Promise<void> {
    await this.plugin.onStop()
  }

  /**
   * Wait blocks until the plugin stops emitting activity for the specified duration.
   */
  async wait(timeoutMs = 100): Promise<void> {
    while (true) {
      const gotActivity = await Promise.race([
        this.waitForActivity(),
        sleep(timeoutMs).then(() => false as const),
      ])
      if (gotActivity === false) return
    }
  }

  async injectSignal(signal: Signal): Promise<this> {
    this.pingActivity()
    await this.plugin.onSignal(this.mock, signal)
    return this
  }

  async injectPayload(payload: Payload): Promise<this> {
    this.pingActivity()
    await this.plugin.onPayload(this.mock, payload)
    return this
  }

  onSignal(cb: (port: number, signal: Signal) => void): this {
    this.mock.onSignalHook = cb
    return this
  }

  onPayload(cb: (port: number, payload: Payload) => void): this {
    this.mock.onPayloadHook = cb
    return this
  }

  // ── private ──────────────────────────────────────────────────────────────

  private pingActivity(): void {
    if (this.activityWaiter) {
      const resolve = this.activityWaiter
      this.activityWaiter = null
      resolve()
      return
    }
    if (this.activityBuffer.length < ACTIVITY_BUFFER_SIZE) {
      this.activityBuffer.push(() => {})
    }
  }

  private waitForActivity(): Promise<true> {
    if (this.activityBuffer.length > 0) {
      this.activityBuffer.shift()
      return Promise.resolve(true)
    }
    return new Promise<true>((resolve) => {
      this.activityWaiter = () => resolve(true)
    })
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}
